dask/array/reductions.py

```diff
@@ -231,7 +231,12 @@ def nanmax(a, axis=None, keepdims=False, split_every=None, out=None):
 
 def numel(x, **kwargs):
     """ A reduction to count the number of elements """
-    return chunk.sum(np.ones_like(x), **kwargs)
+    if hasattr(x, 'todense'):
+        x_ones = x.todense()
+        x_ones[:] = 1
+    else:
+        x_ones = np.ones_like(x)
+    return chunk.sum(x_ones, **kwargs)
```

Comment:
Instead perhaps we replace `np.ones_like` with `np.ones(shape=x.shape, dtype='u1')` ?

Reply:
Indeed :)

Reply:
I amended my commit with your suggestion. It can still allocate a lot of unnecessary memory if the arrays are very sparse and the chunk dimensions comparatively large.

Reply:
Actually using `u1` is wrong. We should probably use `u8` to be able to count large dimensions.

Reply:
Yes, ideally we would re-implement the axis and keepdims logic.  We've been lazy so far.

U1 seems to work for me?

```python
In [1]: import numpy as np

In [2]: np.ones(shape=(1000, 2000), dtype='u1').sum()
Out[2]: 2000000
```

Reply:
There are still broken tests with masked arrays. Need to investigate.